MaterialLib/Component/CreateComponentProperties.cpp

```diff
@@ -0,0 +1,57 @@
+/**
+ * \copyright
+ * Copyright (c) 2012-2018, OpenGeoSys Community (http://www.opengeosys.org)
+ *            Distributed under a Modified BSD License.
+ *              See accompanying file LICENSE.txt or
+ *              http://www.opengeosys.org/project/license
+ *
+ * \file:   CreateComponentProperties.cpp
+ */
+
+#include "CreateComponentProperties.h"
+#include "BaseLib/ConfigTree.h"
+#include "ComponentProperties.h"
+#include "Diffusion/createDiffusion.h"
+#include "Dispersivity/createDispersivity.h"
+
+namespace MaterialLib
+{
+namespace Component
+{
+std::vector<ComponentProperties> createComponentProperties(
+    BaseLib::ConfigTree const& configs,
+    std::vector<std::unique_ptr<ProcessLib::ParameterBase>> const& parameters)
+{
+    DBUG("Create PerComponentProperties.");
+    std::vector<ComponentProperties> per_component_properties;
+
+    auto const& component_configs =
+        //! \ogs_file_param{material__components}
+        configs.getConfigSubtree("components");
+
+    for (auto const& comp_config :
+         //! \ogs_file_param{material__components__component}
+         component_configs.getConfigSubtreeList("component"))
+    {
+        auto name = comp_config.getConfigAttribute<std::string>("name");
+
+        auto const& diffusion_config =
+            //! \ogs_file_param{material__components__component__diffusion}
+            comp_config.getConfigSubtree("diffusion");
+        auto diffusion = MaterialLib::Component::createDiffusion(
+            diffusion_config, parameters);
+
+        auto const& dispersivity_config =
+            //! \ogs_file_param{material__components__component__dispersivity}
+            comp_config.getConfigSubtree("dispersivity");
+        auto dispersivity = MaterialLib::Component::createDispersivity(
+            dispersivity_config, parameters);
+
+        per_component_properties.emplace_back(ComponentProperties{
```

Comment:
Function `emplace_back` constructs the object itself. I think either
```pcp.emplace_back(std::move(name), std::move(diffusion), std::move(dispersivity));```
or
```pcp.push_back(ComponentProperties{std::move(name), std::move(diffusion), std::move(dispersivity)});```

Reply:
Corrected. You are right. I need to take a close look at the difference between emplace_back and push back.